### PR TITLE
Support for TIFF format in Depth Compression

### PR DESCRIPTION
**Pull request by xiangzhi**
This pull request add the support to encode/decode depth image with TIFF compression format. TIFF format has been used in application such as ROS to encode depth images. The code/design for this PR is based on the PNG encoder/decoder.

Changes:
- Add TIFF as an DepthCompressionMethod.
- Added Encoder/Decoder for Windows.
- Added test for TIFF encoding/decoding.

Let me know if any changes are required

### CODE REVIEW

File: Sources/Imaging/Microsoft.Psi.Imaging.Windows/DepthImageFromTiffStreamDecoder.cs [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+namespace Microsoft.Psi.Imaging
+{
+    using System.IO;
+    using System.Windows;
+    using System.Windows.Media.Imaging;
+
+    /// <summary>
+    /// Implements a depth image decoder.
+    /// </summary>
+    public class DepthImageFromTiffStreamDecoder : IDepthImageFromStreamDecoder
```
Comment (chitsaw) [inline Sources/Imaging/Microsoft.Psi.Imaging.Windows/DepthImageFromTiffStreamDecoder.cs:13]: I'm not sure you need to define a TIFF-specific decoder class. If you look at the `DepthImageFromStreamDecoder` class and how that is used, it should be capable of instantiating the correct decoder (e.g. PNG, TIFF) based on the stream. Did you try it (in your test method) and if so did it not work for decoding your TIFF-encoded depth images?

File: Sources/Imaging/Test.Psi.Imaging.Windows/ImageTester.cs [hunk 2]
```diff
@@ -1035,6 +1035,21 @@ public void DepthImage_Serialize()
...
+        public void EncodeDepthImageAsTiff()
+        {
+            // generate a "depth image" for testing
+            var testImage16bpp = new Image(this.testImage.Width, this.testImage.Height, PixelFormat.Gray_16bpp);
+            this.testImage.CopyTo(testImage16bpp);
+            var testDepthImage = DepthImage.CreateFrom(testImage16bpp.ToBitmap());
+
+            // encode to TIFF and decode
+            var encodedDepthTiffImage = testDepthImage.Encode(new DepthImageToTiffStreamEncoder());
+            var decodedDepthImage = encodedDepthTiffImage.Decode(new DepthImageFromTiffStreamDecoder());
```
Comment (chitsaw) [inline Sources/Imaging/Test.Psi.Imaging.Windows/ImageTester.cs:1049]: Try using `DepthImageFromStreamDecoder()` here

File: Sources/Imaging/Microsoft.Psi.Imaging.Windows/DepthImageCompressor.cs [hunk 3]
```diff
@@ -27,12 +27,16 @@ public DepthImageCompressor(DepthCompressionMethod depthCompressionMethod)
             {
                 case DepthCompressionMethod.Png:
                     this.encoder = new DepthImageToPngStreamEncoder();
+                    this.decoder = new DepthImageFromStreamDecoder();
+                    break;
+                case DepthCompressionMethod.Tiff:
+                    this.encoder = new DepthImageToTiffStreamEncoder();
+                    this.decoder = new DepthImageFromTiffStreamDecoder();
```
Comment (chitsaw) [inline Sources/Imaging/Microsoft.Psi.Imaging.Windows/DepthImageCompressor.cs:34]: Should be able to just assign the `this.encoder` and use the generic `DepthImageFromStreamDecoder` per the original code. Unless this somehow didn't work?

File: Sources/Imaging/Microsoft.Psi.Imaging.Linux/DepthImageCompressor.cs [hunk 4]
```diff
@@ -28,6 +28,9 @@ public DepthImageCompressor(DepthCompressionMethod depthCompressionMethod)
                 case DepthCompressionMethod.Png:
                     this.encoder = new DepthImageToPngStreamEncoder();
                     break;
+                case DepthCompressionMethod.Tiff:
+                    // Currently not supported in Linux.
+                    break;
```
Comment (chitsaw) [inline Sources/Imaging/Microsoft.Psi.Imaging.Linux/DepthImageCompressor.cs:33]: I would throw a `NotImplementedException` here to make it clear to the user that it isn't implemented yet.

**Review by chitsaw — CHANGES_REQUESTED** [review]
Thanks for submitting these additions! Please see comments.

File: Sources/Imaging/Microsoft.Psi.Imaging.Windows/DepthImageFromTiffStreamDecoder.cs (hunk 1 above)
Comment (xiangzhi) [reply Sources/Imaging/Microsoft.Psi.Imaging.Windows/DepthImageFromTiffStreamDecoder.cs:13]: You are right, the existing decoder still works for TIFF files. I did not try it before and rolling back the change now.

File: Sources/Imaging/Test.Psi.Imaging.Windows/ImageTester.cs (hunk 2 above)
Comment (xiangzhi) [reply Sources/Imaging/Test.Psi.Imaging.Windows/ImageTester.cs:1049]: Replaced with `DepthImageFromStreamDecoder()`

File: Sources/Imaging/Microsoft.Psi.Imaging.Windows/DepthImageCompressor.cs (hunk 3 above)
Comment (xiangzhi) [reply Sources/Imaging/Microsoft.Psi.Imaging.Windows/DepthImageCompressor.cs:34]: I reverted the change and only added the encoders.

File: Sources/Imaging/Microsoft.Psi.Imaging.Linux/DepthImageCompressor.cs (hunk 4 above)
Comment (xiangzhi) [reply Sources/Imaging/Microsoft.Psi.Imaging.Linux/DepthImageCompressor.cs:33]: Changed to throwing `NotImplementedException`.

**Comment by xiangzhi** [comment]
Thanks for the review. I just made the suggested changes.

**Review by chitsaw — APPROVED** [review]
Thanks for making the changes!

**Pull request merged**